components/servo/lib.rs

```diff
@@ -240,11 +247,228 @@ impl<Window> Servo<Window> where Window: WindowMethods + 'static {
+                    self.compositor.window.set_inner_size(top_level_browsing_context, size);
+                },
+
+                (EmbedderMsg::GetClientWindow(top_level_browsing_context, send),
+                 ShutdownState::NotShuttingDown) => {
+                    let rect = self.compositor.window.client_window(top_level_browsing_context);
+                    if let Err(e) = send.send(rect) {
+                        warn!("Sending response to get client window failed ({}).", e);
+                    }
+                },
+
+                (EmbedderMsg::AllowNavigation(top_level_browsing_context,
+                                              url,
+                                              response_chan),
+                 ShutdownState::NotShuttingDown) => {
+                    self.compositor.window.allow_navigation(top_level_browsing_context, url, response_chan);
+                },
+
+                (EmbedderMsg::KeyEvent(top_level_browsing_context,
+                                       ch,
+                                       key,
+                                       state,
+                                       modified),
+                 ShutdownState::NotShuttingDown) => {
+                    if state == KeyState::Pressed {
+                        self.compositor.window.handle_key(top_level_browsing_context, ch, key, modified);
+                    }
+                },
+
+                (EmbedderMsg::SetCursor(cursor), ShutdownState::NotShuttingDown) => {
+                    self.compositor.window.set_cursor(cursor)
+                },
+
+                (EmbedderMsg::NewFavicon(top_level_browsing_context, url), ShutdownState::NotShuttingDown) => {
+                    self.compositor.window.set_favicon(top_level_browsing_context, url);
+                },
+
+                (EmbedderMsg::HeadParsed(top_level_browsing_context, ), ShutdownState::NotShuttingDown) => {
+                    self.compositor.window.head_parsed(top_level_browsing_context, );
+                },
+
+                (EmbedderMsg::HistoryChanged(top_level_browsing_context, entries, current),
+                 ShutdownState::NotShuttingDown) => {
+                    self.compositor.window.history_changed(top_level_browsing_context, entries, current);
+                },
+
+                (EmbedderMsg::SetFullscreenState(top_level_browsing_context, state),
+                 ShutdownState::NotShuttingDown) => {
+                    self.compositor.window.set_fullscreen_state(top_level_browsing_context, state);
+                },
+
+                (EmbedderMsg::LoadStart(top_level_browsing_context), ShutdownState::NotShuttingDown) => {
+                    self.compositor.window.load_start(top_level_browsing_context);
+                },
+
+                (EmbedderMsg::LoadComplete(top_level_browsing_context), ShutdownState::NotShuttingDown) => {
+                    self.compositor.on_load_complete();
```

Comment:
I'm not sure, the method calls [`recomposite_if_neccessary`](https://github.com/servo/servo/pull/17269/files#diff-f8d10559c852d773193e1ad337448532R654) of compositor. I did add this `on_load_complete` method, which perhaps is in the spirit of putting all compositor logic in compositor(as opposed to checking the `opts` right here?